crates/core/transaction/src/is_action.rs

```diff
@@ -329,10 +329,44 @@ impl IsAction for Swap {
         });
 
         ActionView::Swap(match plaintext {
-            Some(swap_plaintext) => SwapView::Visible {
-                swap: self.to_owned(),
-                swap_plaintext,
-            },
+            Some(swap_plaintext) => {
+                // If we can find a matching BSOD in the TxP, use it to compute the output notes
+                // for the swap.
+                let bsod = txp
+                    .batch_swap_output_data
+                    .iter()
+                    // This finds the first matching one; there should only be one
+                    // per trading pair per block and we trust the TxP provider not to lie about it.
+                    .find(|bsod| bsod.trading_pair == swap_plaintext.trading_pair);
+
+                let (output_1, output_2) = match bsod.map(|bsod| swap_plaintext.output_notes(bsod))
+                {
+                    Some((output_1, output_2)) => {
+                        (Some(txp.view_note(output_1)), Some(txp.view_note(output_2)))
+                    }
+                    None => (None, None),
+                };
+
+                SwapView::Visible {
+                    swap: self.to_owned(),
+                    swap_plaintext: swap_plaintext.clone(),
+                    output_1,
+                    output_2,
+                    claim_tx: txp
+                        .nullification_transaction_ids_by_commitment
+                        .get(&commitment)
+                        .cloned(),
+                    batch_swap_output_data: bsod.cloned(),
+                    asset_1_metadata: txp
+                        .denoms
+                        .get(&swap_plaintext.trading_pair.asset_1())
+                        .cloned(),
+                    asset_2_metadata: txp
+                        .denoms
+                        .get(&swap_plaintext.trading_pair.asset_2())
+                        .cloned(),
```

Comment:
Populating the `asset_1_metadata` and `asset_2_metadata` is important to be able to view the `Swap` action pre-submission, where there won't be any output notes.